### PR TITLE
[sw/silicon_creator] Introduce the sec_mmio module

### PR DESCRIPTION
**Pull request by moidx**
The `sec_mmio` module is responsible for tracking critical register values for an initialized `sec_mmio_ctx_t` context, and provides  a mechanism to check expected values and trigger shutdown escalation on fault detection.

**Initialization**

*    `sec_mmio_init()`.

**Register writes**

*   Perform a number (N) of calls to `sec_mmio_write32()`.
*   Increment the expected number of writes by N by calling `sec_mmio_write_increment()`. This is done using a separate function call to be able to detect skip instruciton faults on `sec_mmio_write32()` calls.

**Register reads**

*   `sec_mmio_read32()`.
 
**Expectation checks**

See the documentation on the following functions:

*   `sec_mmio_check_values()`
*   `sec_mmio_check_counters()`

**Testing**

Test cases can use a mock sec mmio handle inside the test class. The test class must inherit from `mask_rom_test::MaskRomTest`, for example:

```cc
class CriticalDeviceTest : public mask_rom_test::MaskRomTest {
 protected:
  uint32_t base_ = TOP_EARLGREY_CRITICAL_DEVICE_BASE_ADDR;
  mask_rom_test::MockSecMmio mmio_;  // This handle is required.
};
```

Expectations are similar to the dif mock mmio version, with the difference that the mmio_ handle and the base_ address need to be explicitly set inside the macro:

```cc
 EXPECT_SEC_WRITE32(mmio_, base_ + CRITICAL_DEVICE_CFG_REG_OFFSET,
                      {
                          {CRITICAL_DEVICE_CFG_FOO_BIT, false},
                      });
```

**Opens**

*   sec_mmio_ctx is currently defined as an extern to simplify testing. For the actual target, we need to define a memory region to share the data with ROM_EXT.
*   Currently fault detection escalations are performed by calling a handler that is registered at `sec_mmio_init()` call time. Need to determine if we want to move to a mock_shutdown implementation, or if we want to refactor the code to return error codes.

### CODE REVIEW

**Comment by moidx** [comment]
The first commit in this PR is under review in #6464.

File: sw/device/silicon_creator/lib/base/sec_mmio.c
Comment (cfrantz) [inline sw/device/silicon_creator/lib/base/sec_mmio.c:37]: What if `i >= kSecMmioRegFileSize` ?

Is this a condition under which we'd invoke shutdown?

File: sw/device/silicon_creator/lib/base/sec_mmio.h
Comment (cfrantz) [inline sw/device/silicon_creator/lib/base/sec_mmio.h:61]: I assume this is just an initial estimate.  Will we tune this later?

File: sw/device/silicon_creator/lib/base/sec_mmio.h
Comment (cfrantz) [inline sw/device/silicon_creator/lib/base/sec_mmio.h:170]: How do we expect the caller to generate the rnd_offset in the Mask ROM?

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (cfrantz) [inline sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:120]: Are there any tests which use a different rnd_offset?

File: sw/device/silicon_creator/lib/base/sec_mmio.c
Comment (mcy) [inline sw/device/silicon_creator/lib/base/sec_mmio.c:14]: I think I said this in a different version of this CL, but names starting with __ are reserved words in C. I think it's worth picking a different one here...

File: sw/device/silicon_creator/lib/base/sec_mmio.c
Comment (mcy) [inline sw/device/silicon_creator/lib/base/sec_mmio.c:95]: I think it may be useful to figure out if we need a value barrier here. I suspect the function call in the `if` is sufficient, but if we were to repalce it with
```c
__shutdown_handler();
__builtin_unreachable();
```
(which we probably should) this might make the compiler think "ah yes this is only reachable if the condition was already satisfied" and delete the branch altogether.

In particular, I think it is prudent to add the following macro to hardened.h and start using it:
```c
#define OT_VALUE_BARRIER(x) asm volatile ("", "=r"(x))
```
(This is a standard trick for clobbering all compiler-deduced knowledge about `x`, because it cannot peek into the assembly block.)

File: sw/device/silicon_creator/lib/base/sec_mmio.c
Comment (mcy) [inline sw/device/silicon_creator/lib/base/sec_mmio.c:108]: I think the compiler may be entitled to delete this volatile... what optimization are you trying to inhibit, given that kValZero is already "poisoned" in a sense?

File: sw/device/silicon_creator/lib/base/sec_mmio.c
Comment (mcy) [inline sw/device/silicon_creator/lib/base/sec_mmio.c:116]: Maybe specify the precise check here? The combination of four values into all ones is not immediately clear.

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (mcy) [reply sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:120]: Perhaps a test that hammers it with a range of random values is warranted?

File: sw/device/silicon_creator/lib/base/sec_mmio.c
Comment (mcy) [reply sw/device/silicon_creator/lib/base/sec_mmio.c:95]: Someone internal pointed me to https://github.com/google/benchmark/blob/master/include/benchmark/benchmark.h#L328 as another phrasing of this...

Comment (asb) [reply sw/device/silicon_creator/lib/base/sec_mmio.c:95]: I was going to point to the Google benchmark macros too. Thanks for the boringssl link. Ultimately, both the boringssl and Google benchmark approaches were originated by Chandler Carruth who has spent quite some time thinking about this, so we'd be well advised to adopt whichever of these is most appropriate. Do note that the boringssl `value_barrier_*` is weaker than the Google benchmark `DoNotOptimize`.

There's a separate longer discussion to be had about what properties of code like this we want the compiler to maintain and how to ensure that - i.e. are the barriers like the above sufficient? Are we applying them appropriately? Do we have the necessary process in place to review that (and potentially ensure they work as expected as compilers evolve?)? etc etc. This shouldn't block landing the PR of course.

Comment (mcy) [reply sw/device/silicon_creator/lib/base/sec_mmio.c:95]: I'm happy to start a doc on this at some point. I think it should include:
- What the barrier is for, and the extremely limited circumstances under which to use it.
- Justification on the correctness of the barrier implementation; it is unclear what optimizations we want to block, so making a list of those may be appropriate.
- Review criteria for use of the barrier.

(Obv this should not block the PR, I expect there will be a mask_rom-wide audit of this eventually.)

Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio.c:95]: Can you guys start contributing to this doc? 

https://docs.google.com/document/d/184RWslB-9baXy1c5cKDAiXcqzvwSFFGlL8XYaNoks-Y/edit

Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio.c:95]: Filed #6610 to keep track of this open.

File: sw/device/silicon_creator/lib/base/sec_mmio.c
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio.c:37]: Done. Added this check as this serves to enhance fault detection.

File: sw/device/silicon_creator/lib/base/sec_mmio.h
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio.h:64]: Done. Also added an issue to track this.

File: sw/device/silicon_creator/lib/base/sec_mmio.h
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio.h:178]: Added more details to the stringdoc. The caller is expected to use the entropy source or the csrng driver to generate `rnd_offset`.

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:131]: Done. Added additional test cases to test different permutations.

File: sw/device/silicon_creator/lib/base/sec_mmio.c
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio.c]: Done. Updated to `sec_mmio_shutdown_cb`.

File: sw/device/silicon_creator/lib/base/sec_mmio.c
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio.c]: You are right. Removed use of `volatile` here as it seems unnecessary.

**Comment by moidx** [comment]
Hi @mcy, @cfrantz. PTAL. Thanks!

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (alphan) [inline sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:37]: is this used anywhere else?

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (alphan) [inline sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:31]: WDYT about writing the tests that call the shutdown callback as [death tests](https://github.com/google/googletest/blob/master/docs/advanced.md#death-tests)?

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (alphan) [inline sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:42]: just wanted to point out that the global ctx is zero initialized. You may want to initialize it with different values (before init is called).

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (alphan) [inline sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:69]: nit: OR

File: sw/device/silicon_creator/lib/base/sec_mmio.c
Comment (alphan) [inline sw/device/silicon_creator/lib/base/sec_mmio.c:46]: do we need this if the callback is declared noreturn?

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (alphan) [inline sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:142]: Can you clarify what this means?

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (alphan) [inline sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:185]: This seems to test undefined behavior because the control flow reaches __builtin_unreachable, right?

Comment (mcy) [reply sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:185]: Yeah, that should compile down to a `ud2` so I don't think this test should pass at all.

I'm in favor of wrapping `__builtin_unreachable` such that it calls abort() when we're running tests.

**Review by mcy — APPROVED** [review]
We should resolve Alphan's comments about unreachable and death tests before merging, but otherwise LGTM

**Review by cfrantz — APPROVED** [review]

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:185]: I need to look at this to understand why the test is passing.

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc]: removed.

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc:43]: fixed to initialize data with non-zero data for this test case.

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc]: switched to use `std::abort()` and `ASSERT_DEATH`.

File: sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc
Comment (moidx) [reply sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc]: I removed this comment as it is a bit confusing.

**Pull request merged**
